Implement NodeExcludeBalancers to exclude nodes as external loadbalancer

cyclinder opened the pull request:
The label specifies that the node should not be considered as a target for external load-balancers which use nodes as a second hop.

Fixed https://github.com/metallb/metallb/issues/2021

cyclinder commented:
@fedepaol I'm not sure this change is in the right place, Can you check it for me？Thanks

fedepaol commented:
> @fedepaol I'm not sure this change is in the right place, Can you check it for me？Thanks

I'd follow what we do for the networknotavailablecondition:

https://github.com/metallb/metallb/blob/1f0fdf34a8785a2efd8bc73fff7af972a44e788d/speaker/bgp_controller.go#L169

https://github.com/metallb/metallb/blob/1f0fdf34a8785a2efd8bc73fff7af972a44e788d/speaker/layer2_controller.go#L234

should be pretty easy.

Also, please add a couple of e2e tests.

fedepaol commented:
Also also, please fix the commit message

fedepaol commented:
@cyclinder still interested in working on this?

cyclinder commented:
@fedepaol hey，sorry for the delay. yes, but I am sucked in working these days. I'll working on this next weekend.

fedepaol commented:
> @fedepaol hey，sorry for the delay. yes, but I am sucked in working these days. I'll working on this next weekend.

no problem, I know the feeling :-)
Thanks again!

onesb23 commented:
This problem is relevant for me.

cyclinder commented:
I apologize for the delay. I will work on this in the next few days and address the CI failures.

onesb23 commented:
@cyclinder Thanks, I appreciate it 🤝

cyclinder commented:
Sorry for the delay again.

Before adding the `go.work` at the root of the project, it was really hard to add code for e2etests, and it was challenging to run inv e2etest locally. I'm not sure if it's an issue with my local environment or something else. After adding the go.work file as follows, everything went smoothly.

```go
go 1.21.4

use (
	.
	./e2etest
)

```

The previous CI failures were because when creating a cluster with kind, it labels the control-plane node with "node.kubernetes.io/exclude-from-external-load-balancers" by default., and this PR implements the value of this label, causing some specs to fail. 

I would like to manually remove this label after starting the Kind cluster (to avoid affecting existing e2e tests), and then separately add an e2e test specifically for testing this PR. This e2e will include the following steps:

- Create a service with ETP=Local and announce it to the control-plane node.
- Label the control-plane node with label: "" and manually trigger a configuration update.
- Verify if it is re-announced to other nodes.

To avoid impacting other e2e tests, this e2e must run serially. WDYT?

fedepaol commented on e2etest/pkg/frr/validation.go:
```diff
@@ -60,9 +60,23 @@ func RoutesMatchNodes(nodes []v1.Node, route Route, ipFamily ipfamily.Family, vr
 		delete(nodesIPs, h.String())
 	}
+
+	excludesNodeIPs, err := k8s.NodeIPsForExcludeBalancers(nodes, ipFamily, vrfName)
```
I'd rather have this outside (on the calling site), with an utility function that filters the nodes from the []v1.Node slice (and only in the relevant test)

cyclinder replied:
please forget it... :(

fedepaol commented on e2etest/pkg/k8s/nodes.go:
```diff
@@ -59,6 +59,51 @@ func NodeIPsForFamily(nodes []v1.Node, family ipfamily.Family, vrfName string) (
 
+// NodeIPsForExcludeBalancers return the IP addresses of all nodes with the label
+// 'node.kubernetes.io/exclude-from-external-load-balancers'
+func NodeIPsForExcludeBalancers(nodes []v1.Node, family ipfamily.Family, vrfName string) ([]string, error) {
```
As above, it'd be simpler to just filter out the node as a whole (or am I missing something?)

cyclinder replied:
please forget it... :(

fedepaol commented on internal/config/config.go:
```diff
@@ -1140,6 +1140,11 @@ func selectedNodes(nodes []corev1.Node, selectors []metav1.LabelSelector) (map[s
 	res := make(map[string]bool)
 OUTER:
 	for _, node := range nodes {
+		_, ok := node.Labels[corev1.LabelNodeExcludeBalancers]
```
this is the wrong place to add the logic, as the function is "selected nodes".

I'd rather do the filtering in the controller (when we build the resources structure), or on top of config.For and carrying around the filtered node slices.

fedepaol replied:
Or, you can even add the filter in the config controller's predicate.

cyclinder replied:
How about the latest changes? I prefer to do these filters in the speaker's shouldAnnoce, it's simply and we can easily add logs to see what happed.

cyclinder replied:
It works well on my local machine:

```
➜  metallb git:(exclude_nodes) ✗ kubectl describe svc -n l2-6976 external-local-lb | grep nodeAssigned
  Normal  nodeAssigned  24m (x6 over 45m)  metallb-speaker  announcing from node "kind-worker2" with protocol "layer2"

➜  metallb git:(exclude_nodes) ✗ kubectl label nodes kind-worker2 node.kubernetes.io/exclude-from-external-load-balancers=""
node/kind-worker2 labeled
➜  metallb git:(exclude_nodes) ✗ kubectl describe svc -n l2-6976 external-local-lb | grep nodeAssigned                      
  Normal  nodeAssigned  24m (x6 over 45m)  metallb-speaker  announcing from node "kind-worker2" with protocol "layer2"
  Normal  nodeAssigned  3s (x2 over 38m)   metallb-speaker  announcing from node "kind-control-plane" with protocol "layer2"

➜  metallb git:(exclude_nodes) ✗ kubectl label nodes kind-worker2 node.kubernetes.io/exclude-from-external-load-balancers-  
node/kind-worker2 unlabeled
➜  metallb git:(exclude_nodes) ✗ kubectl describe svc -n l2-6976 external-local-lb | grep nodeAssigned                    
  Normal  nodeAssigned  55s (x2 over 39m)  metallb-speaker  announcing from node "kind-control-plane" with protocol "layer2"
  Normal  nodeAssigned  4s (x7 over 46m)   metallb-speaker  announcing from node "kind-worker2" with protocol "layer2"
```

fedepaol commented:
> Sorry for the delay again.
> 
> Before adding the `go.work` at the root of the project, it was really hard to add code for e2etests, and it was challenging to run inv e2etest locally. I'm not sure if it's an issue with my local environment or something else. After adding the go.work file as follows, everything went smoothly.
> 

Weird, I did not have any issues, when opening directly the e2etests folder both with vim and vscode. We have a hardcoded go.work file there to fetch the metallb api.

> ```go
> go 1.21.4
> 
> use (
> 	.
> 	./e2etest
> )
> ```
> 
> The previous CI failures were because when creating a cluster with kind, it labels the control-plane node with "node.kubernetes.io/exclude-from-external-load-balancers" by default., and this PR implements the value of this label, causing some specs to fail.
> 
> I would like to manually remove this label after starting the Kind cluster (to avoid affecting existing e2e tests),

I'd do it as part of inv dev-env

> and then separately add an e2e test specifically for testing this PR. This e2e will include the following steps:
> 
> * Create a service with ETP=Local and announce it to the control-plane node.
> * Label the control-plane node with label: "" and manually trigger a configuration update.
> * Verify if it is re-announced to other nodes.
> 

Not sure I understood. Here's what I'd do:
- With L2 advertisement:
- Announce the service and check which node is being used
- Add hte label, the service should move
- Remove the label, the service should get back to the original node 

> To avoid impacting other e2e tests, this e2e must run serially. WDYT?

All the e2e tests run serially.

fedepaol commented:
I would also add a filter in the node controller

cyclinder commented:
>  * With L2 advertisement:
> * Announce the service and check which node is being used
> * Add hte label, the service should move
> * Remove the label, the service should get back to the original node

Thanks, sorry for my bad words, this is what I want to say.

The current changes are just my initial idea,  I will rework this following your suggestions:

- Remove the label in `inv-env`
- With L2 advertisement: Announce the service and check which node is being used
- Add hte label, the service should move
- Remove the label, the service should get back to the original node

> I would also add a filter in the node controller

if a node has this label, it's somewhat similar to the condition of the node being NotReady, right?

cyclinder commented:
[CI / backward_compatible](https://github.com/metallb/metallb/actions/runs/7469906543/job/20327837312?pr=2073) failed, It seems we should remove the `node.kubernetes.io/exclude-from-external-load-balancers` label when the cluster is created. Otherwise, the backward_compatible CI would be affected by these changes.

Do we need a cherry-pick to v0.12.1? @fedepaol 

the cherry-pick includes the following change:

```
diff --git a/tasks.py b/tasks.py
index b4006d43..44041375 100644
--- a/tasks.py
+++ b/tasks.py
@@ -677,6 +677,10 @@ def e2etest(ctx, name="kind", export=None, kubeconfig=None, system_namespaces="k
     else:
         os.environ['KUBECONFIG'] = kubeconfig
 
+    nodes = run("kind get nodes --name {name}".format(name=name)).stdout.strip().split("\n")
+    for node in nodes:
+        run("{} label nodes {} node.kubernetes.io/exclude-from-external-load-balancers-".format(kubectl_path,node), hide=True)
+
     namespaces = system_namespaces.replace(' ', '').split(',')
     for ns in namespaces:
         run("kubectl -n {} wait --for=condition=Ready --all pods --timeout 300s".format(ns), hide=True)
```

fedepaol commented on tasks.py:
```diff
@@ -842,8 +842,11 @@ def e2etest(ctx, name="kind", export=None, kubeconfig=None, system_namespaces="k
     for ns in namespaces:
         run("{} -n {} wait --for=condition=Ready --all pods --timeout 300s".format(kubectl_path, ns), hide=True)
 
+    nodes = run("kind get nodes --name {name}".format(name=name)).stdout.strip().split("\n")
```
Can you comment what this new addition does (or even move to it's own function with a descriptive name)?

cyclinder replied:
sure, thanks

cyclinder commented on internal/k8s/nodes/nodes.go:
```diff
@@ -25,3 +25,15 @@ func conditionStatus(n *corev1.Node, ct corev1.NodeConditionType) corev1.Conditi
 }
+
+// IsLabelNodeExcludeBalancers returns true if the given node has labeld node.kubernetes.io/exclude-from-external-load-balancers".
+func IsLabelNodeExcludeBalancers(n *corev1.Node) bool {
```
thanks

cyclinder commented on e2etest/l2tests/l2.go:
```diff
@@ -286,6 +286,54 @@ var _ = ginkgo.Describe("L2", func() {
+				err := cs.CoreV1().Services(svc.Namespace).Delete(context.TODO(), svc.Name, metav1.DeleteOptions{})
+				framework.ExpectNoError(err)
+			}()
+			time.Sleep(time.Second)
```
There no need this sleep, we also have eventually later.

fedepaol replied:
Can you remove the sleep?

cyclinder replied:
yeah, but I see the sleep also in https://github.com/metallb/metallb/blob/b42fc17b70323b5fd7d53828e14fcbe29ae5b965/e2etest/l2tests/l2.go#L242

Do we need to remove the sleep?

cyclinder commented:
>  I wouldn't bother cherry picking, just add these lines directly in the CI yaml

But we can't guarantee that `kind` and `kubectl` exist in CI, right?

cyclinder commented:
yes, we just want `kubectl` should be there. 

> or we can download it

This looks a bit cumbersome :(

fedepaol commented on tasks.py:
```diff
@@ -842,8 +842,14 @@ def e2etest(ctx, name="kind", export=None, kubeconfig=None, system_namespaces="k
+       when the cluster is created. and when https://github.com/metallb/metallb/pull/2073 was merged in, which will 
+       affect the currently e2e tests, we should remove the label here.
+    """
+    nodes = removenodeslabel(name)
```
let's call this `dont_exclude_from_lb`

cyclinder replied:
thanks!

fedepaol commented:
Couple of other nits, plus (since we need to retouch it), I noticed the commit message got split. Mind adjusting the len of the subject?

Thanks!

cyclinder commented:
@fedepaol Thanks for the tips, All nits have been fixed.

fedepaol commented on e2etest/l2tests/l2.go:
```diff
@@ -286,6 +286,52 @@ var _ = ginkgo.Describe("L2", func() {
+			}, time.Minute, time.Second).ShouldNot(gomega.Equal(nodeToSet))
+
+			ginkgo.By("removing the NodeExcludeBalancers label of the node")
+			k8s.RemoveLabelFromNode(nodeToSet, corev1.LabelNodeExcludeBalancers, cs)
```
I think we should have an afterEach that clears all the labels, otherwise if this test fails before will affect the status of the cluster and invalidate other tests.

cyclinder replied:
I think adding a defer function is enough, not necessary for all cases.

cyclinder replied:
Do you like the latest change?

```
  k8s.AddLabelToNode(nodeToSet, corev1.LabelNodeExcludeBalancers, "", cs)
  defer func() {
  ginkgo.By("removing the NodeExcludeBalancers label of the node")
  k8s.RemoveLabelFromNode(nodeToSet, corev1.LabelNodeExcludeBalancers, cs)
  }()
```

fedepaol replied:
works for me

fedepaol commented:
Mind having a look at the CI failures?
You should have the artifacts here https://github.com/metallb/metallb/actions/runs/7580058154 iirc

fedepaol commented:
> > Mind having a look at the CI failures?
> 
> Sure! I'm looking this, I think the code should be fine, let's see what happens with CI retries.

yep but if it fails it means that the test is flaky, which is something we don't want :-)

fedepaol commented:
If you check https://github.com/metallb/metallb/actions/runs/7580058154/artifacts/1180482395 you should have all the logs of the test with the configuration and the pod logs at the moment of the failure.

cyclinder commented:
```shell
2024-01-21T11:37:57.062461106Z stdout F {"caller":"bgp_controller.go:170","event":"skipping should announce bgp","ips":["fc00:f853:ccd:e793::9"],"level":"debug","pool":"l2-test","protocol":"bgp","reason":"pool not matching my node","service":"l2-9268/external-local-lb","ts":"2024-01-21T11:37:57Z"}
2024-01-21T11:37:57.062465344Z stdout F {"caller":"layer2_controller.go:123","event":"shouldannounce","ips":["fc00:f853:ccd:e793::9"],"level":"debug","nodes":["kind-control-plane","kind-worker2","kind-worker"],"pool":"l2-test","protocol":"l2","service":"l2-9268/external-local-lb","ts":"2024-01-21T11:37:57Z"}
2024-01-21T11:37:57.062472026Z stdout F {"caller":"main.go:384","event":"serviceAnnounced","ips":["fc00:f853:ccd:e793::9"],"level":"info","msg":"service has IP, announcing","pool":"l2-test","protocol":"layer2","ts":"2024-01-21T11:37:57Z"}
2024-01-21T11:37:57.062476544Z stdout F {"caller":"main.go:346","event":"endUpdate","level":"debug","msg":"end of service update","ts":"2024-01-21T11:37:57Z"}
2024-01-21T11:37:57.062481143Z stdout F {"caller":"service_controller.go:112","controller":"ServiceReconciler","end reconcile":"l2-9268/external-local-lb","level":"info","ts":"2024-01-21T11:37:57Z"}
2024-01-21T11:37:57.201769077Z stdout F {"caller":"config_controller.go:58","controller":"ConfigReconciler","level":"info","start reconcile":"/l2-3382","ts":"2024-01-21T11:37:57Z"}
2024-01-21T11:37:57.201835551Z stdout F {"caller":"config_controller.go:156","controller":"ConfigReconciler","event":"configuration did not change, ignoring","level":"debug","ts":"2024-01-21T11:37:57Z"}
2024-01-21T11:37:57.201840941Z stdout F {"caller":"config_controller.go:157","controller":"ConfigReconciler","end reconcile":"/l2-3382","level":"info","ts":"2024-01-21T11:37:57Z"}
2024-01-21T11:37:57.242166001Z stdout F {"caller":"node_controller.go:46","controller":"NodeReconciler","level":"info","start reconcile":"/kind-worker2","ts":"2024-01-21T11:37:57Z"}
2024-01-21T11:37:57.24220864Z stdout F {"caller":"speakerlist.go:274","level":"info","msg":"triggering discovery","op":"memberDiscovery","ts":"2024-01-21T11:37:57Z"}
2024-01-21T11:37:57.242216724Z stdout F {"caller":"bgp_controller.go:387","event":"nodeLabelsChanged","level":"info","msg":"Node labels changed, resyncing BGP peers","ts":"2024-01-21T11:37:57Z"}
2024-01-21T11:37:57.242221804Z stdout F {"caller":"node_controller.go:62","controller":"NodeReconciler","event":"force service reload","level":"info","ts":"2024-01-21T11:37:57Z"}
2024-01-21T11:37:57.242227064Z stdout F {"caller":"node_controller.go:64","controller":"NodeReconciler","end reconcile":"/kind-worker2","level":"info","ts":"2024-01-21T11:37:57Z"}
2024-01-21T11:37:57.242232744Z stdout F {"caller":"service_controller_reload.go:62","controller":"ServiceReconciler - reprocessAll","level":"info","start reconcile":"metallbreload/reload","ts":"2024-01-21T11:37:57Z"}
2024-01-21T11:37:57.242389244Z stdout F {"caller":"service_controller_reload.go:92","controller":"ServiceReconciler - reprocessAll","level":"debug","reprocessing service":"{\"kind\":\"Service\",\"apiVersion\":\"v1\",\"metadata\":{\"name\":\"external-local-lb\",\"namespace\":\"l2-9268\",\"uid\":\"0fe4f168-a882-4a96-aebd-328f86357e43\",\"resourceVersion\":\"2555\",\"creationTimestamp\":\"2024-01-21T11:37:51Z\",\"labels\":{\"testid\":\"external-local-lb-80aae6b4-b5ab-42cb-9464-2e855f4da42c\"},\"annotations\":{\"metallb.universe.tf/ip-allocated-from-pool\":\"l2-test\"},\"managedFields\":[{\"manager\":\"controller\",\"operation\":\"Update\",\"apiVersion\":\"v1\",\"time\":\"2024-01-21T11:37:51Z\",\"fieldsType\":\"FieldsV1\",\"fieldsV1\":{\"f:metadata\":{\"f:annotations\":{\".\":{},\"f:metallb.universe.tf/ip-allocated-from-pool\":{}}},\"f:status\":{\"f:loadBalancer\":{\"f:ingress\":{}}}},\"subresource\":\"status\"},{\"manager\":\"e2etest.test\",\"operation\":\"Update\",\"apiVersion\":\"v1\",\"time\":\"2024-01-21T11:37:51Z\",\"fieldsType\":\"FieldsV1\",\"fieldsV1\":{\"f:metadata\":{\"f:labels\":{\".\":{},\"f:testid\":{}}},\"f:spec\":{\"f:allocateLoadBalancerNodePorts\":{},\"f:externalTrafficPolicy\":{},\"f:internalTrafficPolicy\":{},\"f:ports\":{\".\":{},\"k:{\\\"port\\\":80,\\\"protocol\\\":\\\"TCP\\\"}\":{\".\":{},\"f:port\":{},\"f:protocol\":{},\"f:targetPort\":{}}},\"f:selector\":{},\"f:sessionAffinity\":{},\"f:type\":{}}}}]},\"spec\":{\"ports\":[{\"protocol\":\"TCP\",\"port\":80,\"targetPort\":80,\"nodePort\":31137}],\"selector\":{\"testid\":\"external-local-lb-80aae6b4-b5ab-42cb-9464-2e855f4da42c\"},\"clusterIP\":\"fd00:10:96::c655\",\"clusterIPs\":[\"fd00:10:96::c655\"],\"type\":\"LoadBalancer\",\"sessionAffinity\":\"None\",\"externalTrafficPolicy\":\"Cluster\",\"ipFamilies\":[\"IPv6\"],\"ipFamilyPolicy\":\"SingleStack\",\"allocateLoadBalancerNodePorts\":true,\"internalTrafficPolicy\":\"Cluster\"},\"status\":{\"loadBalancer\":{\"ingress\":[{\"ip\":\"fc00:f853:ccd:e793::9\"}]}}}","ts":"2024-01-21T11:37:57Z"}
2024-01-21T11:37:57.242477527Z stdout F {"caller":"main.go:297","event":"startUpdate","level":"debug","msg":"start of service update","ts":"2024-01-21T11:37:57Z"}
2024-01-21T11:37:57.24248399Z stdout F {"caller":"bgp_controller.go:170","event":"skipping should announce bgp","ips":["fc00:f853:ccd:e793::9"],"level":"debug","pool":"l2-test","protocol":"bgp","reason":"pool not matching my node","service":"l2-9268/external-local-lb","ts":"2024-01-21T11:37:57Z"}
2024-01-21T11:37:57.242492846Z stdout F {"caller":"layer2_controller.go:123","event":"shouldannounce","ips":["fc00:f853:ccd:e793::9"],"level":"debug","nodes":["kind-control-plane","kind-worker"],"pool":"l2-test","protocol":"l2","service":"l2-9268/external-local-lb","ts":"2024-01-21T11:37:57Z"}
2024-01-21T11:37:57.242551024Z stdout F {"caller":"main.go:428","event":"serviceWithdrawn","ip":["fc00:f853:ccd:e793::9"],"ips":["fc00:f853:ccd:e793::9"],"level":"info","msg":"withdrawing service announcement","pool":"l2-test","protocol":"layer2","reason":"notOwner","ts":"2024-01-21T11:37:57Z"}
2024-01-21T11:37:57.242559339Z stdout F {"caller":"main.go:346","event":"endUpdate","level":"debug","msg":"end of service update","ts":"2024-01-21T11:37:57Z"}
2024-01-21T11:37:57.242696974Z stdout F {"caller":"service_controller_reload.go:92","controller":"ServiceReconciler - reprocessAll","level":"debug","reprocessing service":"{\"kind\":\"Service\",\"apiVersion\":\"v1\",\"metadata\":{\"name\":\"prometheus-operated\",\"namespace\":\"monitoring\",\"uid\":\"e0266f13-56f7-4818-b0ff-840c653af2ec\",\"resourceVersion\":\"651\",\"creationTimestamp\":\"2024-01-21T11:28:02Z\",\"labels\":{\"operated-prometheus\":\"true\"},\"ownerReferences\":[{\"apiVersion\":\"monitoring.coreos.com/v1\",\"kind\":\"Prometheus\",\"name\":\"k8s\",\"uid\":\"3ecabe9a-7e3c-43ee-8db8-669747938f21\"}],\"managedFields\":[{\"manager\":\"PrometheusOperator\",\"operation\":\"Update\",\"apiVersion\":\"v1\",\"time\":\"2024-01-21T11:28:02Z\",\"fieldsType\":\"FieldsV1\",\"fieldsV1\":{\"f:metadata\":{\"f:labels\":{\".\":{},\"f:operated-prometheus\":{}},\"f:ownerReferences\":{\".\":{},\"k:{\\\"uid\\\":\\\"3ecabe9a-7e3c-43ee-8db8-669747938f21\\\"}\":{}}},\"f:spec\":{\"f:clusterIP\":{},\"f:internalTrafficPolicy\":{},\"f:ports\":{\".\":{},\"k:{\\\"port\\\":9090,\\\"protocol\\\":\\\"TCP\\\"}\":{\".\":{},\"f:name\":{},\"f:port\":{},\"f:protocol\":{},\"f:targetPort\":{}}},\"f:selector\":{},\"f:sessionAffinity\":{},\"f:type\":{}}}}]},\"spec\":{\"ports\":[{\"name\":\"web\",\"protocol\":\"TCP\",\"port\":9090,\"targetPort\":\"web\"}],\"selector\":{\"app.kubernetes.io/name\":\"prometheus\"},\"clusterIP\":\"None\",\"clusterIPs\":[\"None\"],\"type\":\"ClusterIP\",\"sessionAffinity\":\"None\",\"ipFamilies\":[\"IPv6\"],\"ipFamilyPolicy\":\"SingleStack\",\"internalTrafficPolicy\":\"Cluster\"},\"status\":{\"loadBalancer\":{}}}","ts":"2024-01-21T11:37:57Z"}
2024-01-21T11:37:57.242842083Z stdout F {"caller":"service_controller_reload.go:92","controller":"ServiceReconciler - reprocessAll","level":"debug","reprocessing service":"{\"kind\":\"Service\",\"apiVersion\":\"v1\",\"metadata\":{\"name\":\"prometheus-operator\",\"namespace\":\"monitoring\",\"uid\":\"dcc16eb4-23e5-4331-96fc-bc0fdb92156e\",\"resourceVersion\":\"574\",\"creationTimestamp\":\"2024-01-21T11:27:55Z\",\"labels\":{\"app.kubernetes.io/component\":\"controller\",\"app.kubernetes.io/name\":\"prometheus-operator\",\"app.kubernetes.io/part-of\":\"kube-prometheus\",\"app.kubernetes.io/version\":\"0.57.0\"},\"managedFields\":[{\"manager\":\"kubectl\",\"operation\":\"Apply\",\"apiVersion\":\"v1\",\"time\":\"2024-01-21T11:27:55Z\",\"fieldsType\":\"FieldsV1\",\"fieldsV1\":{\"f:metadata\":{\"f:labels\":{\"f:app.kubernetes.io/component\":{},\"f:app.kubernetes.io/name\":{},\"f:app.kubernetes.io/part-of\":{},\"f:app.kubernetes.io/version\":{}}},\"f:spec\":{\"f:clusterIP\":{},\"f:ports\":{\"k:{\\\"port\\\":8443,\\\"protocol\\\":\\\"TCP\\\"}\":{\".\":{},\"f:name\":{},\"f:port\":{},\"f:targetPort\":{}}},\"f:selector\":{}}}}]},\"spec\":{\"ports\":[{\"name\":\"https\",\"protocol\":\"TCP\",\"port\":8443,\"targetPort\":\"https\"}],\"selector\":{\"app.kubernetes.io/component\":\"controller\",\"app.kubernetes.io/name\":\"prometheus-operator\",\"app.kubernetes.io/part-of\":\"kube-prometheus\"},\"clusterIP\":\"None\",\"clusterIPs\":[\"None\"],\"type\":\"ClusterIP\",\"sessionAffinity\":\"None\",\"ipFamilies\":[\"IPv6\"],\"ipFamilyPolicy\":\"SingleStack\",\"internalTrafficPolicy\":\"Cluster\"},\"status\":{\"loadBalancer\":{}}}","ts":"2024-01-21T11:37:57Z"}
2024-01-21T11:37:57.242851089Z stdout F {"caller":"service_controller_reload.go:92","controller":"ServiceReconciler - reprocessAll","level":"debug","reprocessing service":"{\"kind\":\"Service\",\"apiVersion\":\"v1\",\"metadata\":{\"name\":\"controller-monitor-service\",\"namespace\":\"metallb-system\",\"uid\":\"86b4a8f9-2e9a-440f-a7ef-2d7fa6363352\",\"resourceVersion\":\"821\",\"creationTimestamp\":\"2024-01-21T11:28:21Z\",\"labels\":{\"name\":\"controller-monitor-service\"},\"annotations\":{\"kubectl.kubernetes.io/last-applied-configuration\":\"{\\\"apiVersion\\\":\\\"v1\\\",\\\"kind\\\":\\\"Service\\\",\\\"metadata\\\":{\\\"annotations\\\":{\\\"prometheus.io/scrape\\\":\\\"true\\\"},\\\"labels\\\":{\\\"name\\\":\\\"controller-monitor-service\\\"},\\\"name\\\":\\\"controller-monitor-service\\\",\\\"namespace\\\":\\\"metallb-system\\\"},\\\"spec\\\":{\\\"clusterIP\\\":\\\"None\\\",\\\"ports\\\":[{\\\"name\\\":\\\"metricshttps\\\",\\\"port\\\":9120,\\\"targetPort\\\":9120}],\\\"selector\\\":{\\\"component\\\":\\\"controller\\\"},\\\"sessionAffinity\\\":\\\"None\\\",\\\"type\\\":\\\"ClusterIP\\\"}}\\n\",\"prometheus.io/scrape\":\"true\"},\"managedFields\":[{\"manager\":\"kubectl-client-side-apply\",\"operation\":\"Update\",\"apiVersion\":\"v1\",\"time\":\"2024-01-21T11:28:21Z\",\"fieldsType\":\"FieldsV1\",\"fieldsV1\":{\"f:metadata\":{\"f:annotations\":{\".\":{},\"f:kubectl.kubernetes.io/last-applied-configuration\":{},\"f:prometheus.io/scrape\":{}},\"f:labels\":{\".\":{},\"f:name\":{}}},\"f:spec\":{\"f:clusterIP\":{},\"f:internalTrafficPolicy\":{},\"f:ports\":{\".\":{},\"k:{\\\"port\\\":9120,\\\"protocol\\\":\\\"TCP\\\"}\":{\".\":{},\"f:name\":{},\"f:port\":{},\"f:protocol\":{},\"f:targetPort\":{}}},\"f:selector\":{},\"f:sessionAffinity\":{},\"f:type\":{}}}}]},\"spec\":{\"ports\":[{\"name\":\"metricshttps\",\"protocol\":\"TCP\",\"port\":9120,\"targetPort\":9120}],\"selector\":{\"component\":\"controller\"},\"clusterIP\":\"None\",\"clusterIPs\":[\"None\"],\"type\":\"ClusterIP\",\"sessionAffinity\":\"None\",\"ipFamilies\":[\"IPv6\"],\"ipFamilyPolicy\":\"SingleStack\",\"internalTrafficPolicy\":\"Cluster\"},\"status\":{\"loadBalancer\":{}}}","ts":"2024-01-21T11:37:57Z"}
2024-01-21T11:37:57.242867139Z stdout F {"caller":"service_controller_reload.go:92","controller":"ServiceReconciler - reprocessAll","level":"debug","reprocessing service":"{\"kind\":\"Service\",\"apiVersion\":\"v1\",\"metadata\":{\"name\":\"prometheus-k8s\",\"namespace\":\"monitoring\",\"uid\":\"f8a0c05b-91a3-4de4-801d-fb7b4a6845ee\",\"resourceVersion\":\"625\",\"creationTimestamp\":\"2024-01-21T11:27:57Z\",\"labels\":{\"app.kubernetes.io/component\":\"prometheus\",\"app.kubernetes.io/instance\":\"k8s\",\"app.kubernetes.io/name\":\"prometheus\",\"app.kubernetes.io/part-of\":\"kube-prometheus\",\"app.kubernetes.io/version\":\"2.36.1\"},\"annotations\":{\"kubectl.kubernetes.io/last-applied-configuration\":\"{\\\"apiVersion\\\":\\\"v1\\\",\\\"kind\\\":\\\"Service\\\",\\\"metadata\\\":{\\\"annotations\\\":{},\\\"labels\\\":{\\\"app.kubernetes.io/component\\\":\\\"prometheus\\\",\\\"app.kubernetes.io/instance\\\":\\\"k8s\\\",\\\"app.kubernetes.io/name\\\":\\\"prometheus\\\",\\\"app.kubernetes.io/part-of\\\":\\\"kube-prometheus\\\",\\\"app.kubernetes.io/version\\\":\\\"2.36.1\\\"},\\\"name\\\":\\\"prometheus-k8s\\\",\\\"namespace\\\":\\\"monitoring\\\"},\\\"spec\\\":{\\\"ports\\\":[{\\\"name\\\":\\\"web\\\",\\\"port\\\":9090,\\\"targetPort\\\":\\\"web\\\"},{\\\"name\\\":\\\"reloader-web\\\",\\\"port\\\":8080,\\\"targetPort\\\":\\\"reloader-web\\\"}],\\\"selector\\\":{\\\"app.kubernetes.io/component\\\":\\\"prometheus\\\",\\\"app.kubernetes.io/instance\\\":\\\"k8s\\\",\\\"app.kubernetes.io/name\\\":\\\"prometheus\\\",\\\"app.kubernetes.io/part-of\\\":\\\"kube-prometheus\\\"},\\\"sessionAffinity\\\":\\\"ClientIP\\\"}}\\n\"},\"managedFields\":[{\"manager\":\"kubectl-client-side-apply\",\"operation\":\"Update\",\"apiVersion\":\"v1\",\"time\":\"2024-01-21T11:27:57Z\",\"fieldsType\":\"FieldsV1\",\"fieldsV1\":{\"f:metadata\":{\"f:annotations\":{\".\":{},\"f:kubectl.kubernetes.io/last-applied-configuration\":{}},\"f:labels\":{\".\":{},\"f:app.kubernetes.io/component\":{},\"f:app.kubernetes.io/instance\":{},\"f:app.kubernetes.io/name\":{},\"f:app.kubernetes.io/part-of\":{},\"f:app.kubernetes.io/version\":{}}},\"f:spec\":{\"f:internalTrafficPolicy\":{},\"f:ports\":{\".\":{},\"k:{\\\"port\\\":8080,\\\"protocol\\\":\\\"TCP\\\"}\":{\".\":{},\"f:name\":{},\"f:port\":{},\"f:protocol\":{},\"f:targetPort\":{}},\"k:{\\\"port\\\":9090,\\\"protocol\\\":\\\"TCP\\\"}\":{\".\":{},\"f:name\":{},\"f:port\":{},\"f:protocol\":{},\"f:targetPort\":{}}},\"f:selector\":{},\"f:sessionAffinity\":{},\"f:sessionAffinityConfig\":{\".\":{},\"f:clientIP\":{\".\":{},\"f:timeoutSeconds\":{}}},\"f:type\":{}}}}]},\"spec\":{\"ports\":[{\"name\":\"web\",\"protocol\":\"TCP\",\"port\":9090,\"targetPort\":\"web\"},{\"name\":\"reloader-web\",\"protocol\":\"TCP\",\"port\":8080,\"targetPort\":\"reloader-web\"}],\"selector\":{\"app.kubernetes.io/component\":\"prometheus\",\"app.kubernetes.io/instance\":\"k8s\",\"app.kubernetes.io/name\":\"prometheus\",\"app.kubernetes.io/part-of\":\"kube-prometheus\"},\"clusterIP\":\"fd00:10:96::f935\",\"clusterIPs\":[\"fd00:10:96::f935\"],\"type\":\"ClusterIP\",\"sessionAffinity\":\"ClientIP\",\"sessionAffinityConfig\":{\"clientIP\":{\"timeoutSeconds\":10800}},\"ipFamilies\":[\"IPv6\"],\"ipFamilyPolicy\":\"SingleStack\",\"internalTrafficPolicy\":\"Cluster\"},\"status\":{\"loadBalancer\":{}}}","ts":"2024-01-21T11:37:57Z"}
2024-01-21T11:37:57.242931248Z stdout F {"caller":"service_controller_reload.go:92","controller":"ServiceReconciler - reprocessAll","level":"debug","reprocessing service":"{\"kind\":\"Service\",\"apiVersion\":\"v1\",\"metadata\":{\"name\":\"kubelet\",\"namespace\":\"kube-system\",\"uid\":\"6d6545e5-147d-41b2-918c-453015604f68\",\"resourceVersion\":\"644\",\"creationTimestamp\":\"2024-01-21T11:28:01Z\",\"labels\":{\"app.kubernetes.io/managed-by\":\"prometheus-operator\",\"app.kubernetes.io/name\":\"kubelet\",\"k8s-app\":\"kubelet\"},\"managedFields\":[{\"manager\":\"PrometheusOperator\",\"operation\":\"Update\",\"apiVersion\":\"v1\",\"time\":\"2024-01-21T11:28:01Z\",\"fieldsType\":\"FieldsV1\",\"fieldsV1\":{\"f:metadata\":{\"f:labels\":{\".\":{},\"f:app.kubernetes.io/managed-by\":{},\"f:app.kubernetes.io/name\":{},\"f:k8s-app\":{}}},\"f:spec\":{\"f:clusterIP\":{},\"f:internalTrafficPolicy\":{},\"f:ports\":{\".\":{},\"k:{\\\"port\\\":4194,\\\"protocol\\\":\\\"TCP\\\"}\":{\".\":{},\"f:name\":{},\"f:port\":{},\"f:protocol\":{},\"f:targetPort\":{}},\"k:{\\\"port\\\":10250,\\\"protocol\\\":\\\"TCP\\\"}\":{\".\":{},\"f:name\":{},\"f:port\":{},\"f:protocol\":{},\"f:targetPort\":{}},\"k:{\\\"port\\\":10255,\\\"protocol\\\":\\\"TCP\\\"}\":{\".\":{},\"f:name\":{},\"f:port\":{},\"f:protocol\":{},\"f:targetPort\":{}}},\"f:sessionAffinity\":{},\"f:type\":{}}}}]},\"spec\":{\"ports\":[{\"name\":\"https-metrics\",\"protocol\":\"TCP\",\"port\":10250,\"targetPort\":10250},{\"name\":\"http-metrics\",\"protocol\":\"TCP\",\"port\":10255,\"targetPort\":10255},{\"name\":\"cadvisor\",\"protocol\":\"TCP\",\"port\":4194,\"targetPort\":4194}],\"clusterIP\":\"None\",\"clusterIPs\":[\"None\"],\"type\":\"ClusterIP\",\"sessionAffinity\":\"None\",\"ipFamilies\":[\"IPv6\",\"IPv4\"],\"ipFamilyPolicy\":\"RequireDualStack\",\"internalTrafficPolicy\":\"Cluster\"},\"status\":{\"loadBalancer\":{}}}","ts":"2024-01-21T11:37:57Z"}
2024-01-21T11:37:57.242982623Z stdout F {"caller":"service_controller_reload.go:92","controller":"ServiceReconciler - reprocessAll","level":"debug","reprocessing service":"{\"kind\":\"Service\",\"apiVersion\":\"v1\",\"metadata\":{\"name\":\"frr-k8s-metrics-service\",\"namespace\":\"metallb-system\",\"uid\":\"eda799d4-9e67-471a-9a74-0ff3ef48f061\",\"resourceVersion\":\"825\",\"creationTimestamp\":\"2024-01-21T11:28:21Z\",\"labels\":{\"app.kubernetes.io/component\":\"kube-rbac-proxy\",\"app.kubernetes.io/created-by\":\"frr-k8s\",\"app.kubernetes.io/instance\":\"frr-k8s-metrics-service\",\"app.kubernetes.io/managed-by\":\"kustomize\",\"app.kubernetes.io/name\":\"service\",\"app.kubernetes.io/part-of\":\"frr-k8s\",\"control-plane\":\"frr-k8s\"},\"annotations\":{\"kubectl.kubernetes.io/last-applied-configuration\":\"{\\\"apiVersion\\\":\\\"v1\\\",\\\"kind\\\":\\\"Service\\\",\\\"metadata\\\":{\\\"annotations\\\":{},\\\"labels\\\":{\\\"app.kubernetes.io/component\\\":\\\"kube-rbac-proxy\\\",\\\"app.kubernetes.io/created-by\\\":\\\"frr-k8s\\\",\\\"app.kubernetes.io/instance\\\":\\\"frr-k8s-metrics-service\\\",\\\"app.kubernetes.io/managed-by\\\":\\\"kustomize\\\",\\\"app.kubernetes.io/name\\\":\\\"service\\\",\\\"app.kubernetes.io/part-of\\\":\\\"frr-k8s\\\",\\\"control-plane\\\":\\\"frr-k8s\\\"},\\\"name\\\":\\\"frr-k8s-metrics-service\\\",\\\"namespace\\\":\\\"metallb-system\\\"},\\\"spec\\\":{\\\"ports\\\":[{\\\"name\\\":\\\"metricshttps\\\",\\\"port\\\":9140,\\\"targetPort\\\":\\\"metricshttps\\\"},{\\\"name\\\":\\\"frrmetricshttps\\\",\\\"port\\\":9141,\\\"targetPort\\\":\\\"frrmetricshttps\\\"}],\\\"selector\\\":{\\\"control-plane\\\":\\\"frr-k8s\\\"}}}\\n\"},\"managedFields\":[{\"manager\":\"kubectl-client-side-apply\",\"operation\":\"Update\",\"apiVersion\":\"v1\",\"time\":\"2024-01-21T11:28:21Z\",\"fieldsType\":\"FieldsV1\",\"fieldsV1\":{\"f:metadata\":{\"f:annotations\":{\".\":{},\"f:kubectl.kubernetes.io/last-applied-configuration\":{}},\"f:labels\":{\".\":{},\"f:app.kubernetes.io/component\":{},\"f:app.kubernetes.io/created-by\":{},\"f:app.kubernetes.io/instance\":{},\"f:app.kubernetes.io/managed-by\":{},\"f:app.kubernetes.io/name\":{},\"f:app.kubernetes.io/part-of\":{},\"f:control-plane\":{}}},\"f:spec\":{\"f:internalTrafficPolicy\":{},\"f:ports\":{\".\":{},\"k:{\\\"port\\\":9140,\\\"protocol\\\":\\\"TCP\\\"}\":{\".\":{},\"f:name\":{},\"f:port\":{},\"f:protocol\":{},\"f:targetPort\":{}},\"k:{\\\"port\\\":9141,\\\"protocol\\\":\\\"TCP\\\"}\":{\".\":{},\"f:name\":{},\"f:port\":{},\"f:protocol\":{},\"f:targetPort\":{}}},\"f:selector\":{},\"f:sessionAffinity\":{},\"f:type\":{}}}}]},\"spec\":{\"ports\":[{\"name\":\"metricshttps\",\"protocol\":\"TCP\",\"port\":9140,\"targetPort\":\"metricshttps\"},{\"name\":\"frrmetricshttps\",\"protocol\":\"TCP\",\"port\":9141,\"targetPort\":\"frrmetricshttps\"}],\"selector\":{\"control-plane\":\"frr-k8s\"},\"clusterIP\":\"fd00:10:96::1382\",\"clusterIPs\":[\"fd00:10:96::1382\"],\"type\":\"ClusterIP\",\"sessionAffinity\":\"None\",\"ipFamilies\":[\"IPv6\"],\"ipFamilyPolicy\":\"SingleStack\",\"internalTrafficPolicy\":\"Cluster\"},\"status\":{\"loadBalancer\":{}}}","ts":"2024-01-21T11:37:57Z"}
2024-01-21T11:37:57.24305153Z stdout F {"caller":"service_controller_reload.go:92","controller":"ServiceReconciler - reprocessAll","level":"debug","reprocessing service":"{\"kind\":\"Service\",\"apiVersion\":\"v1\",\"metadata\":{\"name\":\"frr-k8s-webhook-service\",\"namespace\":\"metallb-system\",\"uid\":\"2748dc7c-c383-4193-a6d6-79f3d7c39952\",\"resourceVersion\":\"829\",\"creationTimestamp\":\"2024-01-21T11:28:21Z\",\"annotations\":{\"kubectl.kubernetes.io/last-applied-configuration\":\"{\\\"apiVersion\\\":\\\"v1\\\",\\\"kind\\\":\\\"Service\\\",\\\"metadata\\\":{\\\"annotations\\\":{},\\\"name\\\":\\\"frr-k8s-webhook-service\\\",\\\"namespace\\\":\\\"metallb-system\\\"},\\\"spec\\\":{\\\"ports\\\":[{\\\"port\\\":443,\\\"targetPort\\\":9443}],\\\"selector\\\":{\\\"control-plane\\\":\\\"webhook-server\\\"}}}\\n\"},\"managedFields\":[{\"manager\":\"kubectl-client-side-apply\",\"operation\":\"Update\",\"apiVersion\":\"v1\",\"time\":\"2024-01-21T11:28:21Z\",\"fieldsType\":\"FieldsV1\",\"fieldsV1\":{\"f:metadata\":{\"f:annotations\":{\".\":{},\"f:kubectl.kubernetes.io/last-applied-configuration\":{}}},\"f:spec\":{\"f:internalTrafficPolicy\":{},\"f:ports\":{\".\":{},\"k:{\\\"port\\\":443,\\\"protocol\\\":\\\"TCP\\\"}\":{\".\":{},\"f:port\":{},\"f:protocol\":{},\"f:targetPort\":{}}},\"f:selector\":{},\"f:sessionAffinity\":{},\"f:type\":{}}}}]},\"spec\":{\"ports\":[{\"protocol\":\"TCP\",\"port\":443,\"targetPort\":9443}],\"selector\":{\"control-plane\":\"webhook-server\"},\"clusterIP\":\"fd00:10:96::6098\",\"clusterIPs\":[\"fd00:10:96::6098\"],\"type\":\"ClusterIP\",\"sessionAffinity\":\"None\",\"ipFamilies\":[\"IPv6\"],\"ipFamilyPolicy\":\"SingleStack\",\"internalTrafficPolicy\":\"Cluster\"},\"status\":{\"loadBalancer\":{}}}","ts":"2024-01-21T11:37:57Z"}
2024-01-21T11:37:57.24312117Z stdout F {"caller":"service_controller_reload.go:92","controller":"ServiceReconciler - reprocessAll","level":"debug","reprocessing service":"{\"kind\":\"Service\",\"apiVersion\":\"v1\",\"metadata\":{\"name\":\"speaker-monitor-service\",\"namespace\":\"metallb-system\",\"uid\":\"080dab07-1a90-4b45-85c5-a9eaece62db1\",\"resourceVersion\":\"832\",\"creationTimestamp\":\"2024-01-21T11:28:21Z\",\"labels\":{\"name\":\"speaker-monitor-service\"},\"annotations\":{\"kubectl.kubernetes.io/last-applied-configuration\":\"{\\\"apiVersion\\\":\\\"v1\\\",\\\"kind\\\":\\\"Service\\\",\\\"metadata\\\":{\\\"annotations\\\":{\\\"prometheus.io/scrape\\\":\\\"true\\\"},\\\"labels\\\":{\\\"name\\\":\\\"speaker-monitor-service\\\"},\\\"name\\\":\\\"speaker-monitor-service\\\",\\\"namespace\\\":\\\"metallb-system\\\"},\\\"spec\\\":{\\\"clusterIP\\\":\\\"None\\\",\\\"ports\\\":[{\\\"name\\\":\\\"metricshttps\\\",\\\"port\\\":9120,\\\"targetPort\\\":9120}],\\\"selector\\\":{\\\"component\\\":\\\"speaker\\\"},\\\"sessionAffinity\\\":\\\"None\\\",\\\"type\\\":\\\"ClusterIP\\\"}}\\n\",\"prometheus.io/scrape\":\"true\"},\"managedFields\":[{\"manager\":\"kubectl-client-side-apply\",\"operation\":\"Update\",\"apiVersion\":\"v1\",\"time\":\"2024-01-21T11:28:21Z\",\"fieldsType\":\"FieldsV1\",\"fieldsV1\":{\"f:metadata\":{\"f:annotations\":{\".\":{},\"f:kubectl.kubernetes.io/last-applied-configuration\":{},\"f:prometheus.io/scrape\":{}},\"f:labels\":{\".\":{},\"f:name\":{}}},\"f:spec\":{\"f:clusterIP\":{},\"f:internalTrafficPolicy\":{},\"f:ports\":{\".\":{},\"k:{\\\"port\\\":9120,\\\"protocol\\\":\\\"TCP\\\"}\":{\".\":{},\"f:name\":{},\"f:port\":{},\"f:protocol\":{},\"f:targetPort\":{}}},\"f:selector\":{},\"f:sessionAffinity\":{},\"f:type\":{}}}}]},\"spec\":{\"ports\":[{\"name\":\"metricshttps\",\"protocol\":\"TCP\",\"port\":9120,\"targetPort\":9120}],\"selector\":{\"component\":\"speaker\"},\"clusterIP\":\"None\",\"clusterIPs\":[\"None\"],\"type\":\"ClusterIP\",\"sessionAffinity\":\"None\",\"ipFamilies\":[\"IPv6\"],\"ipFamilyPolicy\":\"SingleStack\",\"internalTrafficPolicy\":\"Cluster\"},\"status\":{\"loadBalancer\":{}}}","ts":"2024-01-21T11:37:57Z"}
2024-01-21T11:37:57.243182613Z stdout F {"caller":"service_controller_reload.go:92","controller":"ServiceReconciler - reprocessAll","level":"debug","reprocessing service":"{\"kind\":\"Service\",\"apiVersion\":\"v1\",\"metadata\":{\"name\":\"webhook-service\",\"namespace\":\"metallb-system\",\"uid\":\"cc7a48df-88a8-4c03-9c99-5f85d8fd06b4\",\"resourceVersion\":\"836\",\"creationTimestamp\":\"2024-01-21T11:28:21Z\",\"annotations\":{\"kubectl.kubernetes.io/last-applied-configuration\":\"{\\\"apiVersion\\\":\\\"v1\\\",\\\"kind\\\":\\\"Service\\\",\\\"metadata\\\":{\\\"annotations\\\":{},\\\"name\\\":\\\"webhook-service\\\",\\\"namespace\\\":\\\"metallb-system\\\"},\\\"spec\\\":{\\\"ports\\\":[{\\\"port\\\":443,\\\"targetPort\\\":9443}],\\\"selector\\\":{\\\"component\\\":\\\"controller\\\"}}}\\n\"},\"managedFields\":[{\"manager\":\"kubectl-client-side-apply\",\"operation\":\"Update\",\"apiVersion\":\"v1\",\"time\":\"2024-01-21T11:28:21Z\",\"fieldsType\":\"FieldsV1\",\"fieldsV1\":{\"f:metadata\":{\"f:annotations\":{\".\":{},\"f:kubectl.kubernetes.io/last-applied-configuration\":{}}},\"f:spec\":{\"f:internalTrafficPolicy\":{},\"f:ports\":{\".\":{},\"k:{\\\"port\\\":443,\\\"protocol\\\":\\\"TCP\\\"}\":{\".\":{},\"f:port\":{},\"f:protocol\":{},\"f:targetPort\":{}}},\"f:selector\":{},\"f:sessionAffinity\":{},\"f:type\":{}}}}]},\"spec\":{\"ports\":[{\"protocol\":\"TCP\",\"port\":443,\"targetPort\":9443}],\"selector\":{\"component\":\"controller\"},\"clusterIP\":\"fd00:10:96::eb57\",\"clusterIPs\":[\"fd00:10:96::eb57\"],\"type\":\"ClusterIP\",\"sessionAffinity\":\"None\",\"ipFamilies\":[\"IPv6\"],\"ipFamilyPolicy\":\"SingleStack\",\"internalTrafficPolicy\":\"Cluster\"},\"status\":{\"loadBalancer\":{}}}","ts":"2024-01-21T11:37:57Z"}
2024-01-21T11:37:57.243217007Z stdout F {"caller":"service_controller_reload.go:92","controller":"ServiceReconciler - reprocessAll","level":"debug","reprocessing service":"{\"kind\":\"Service\",\"apiVersion\":\"v1\",\"metadata\":{\"name\":\"kubernetes\",\"namespace\":\"default\",\"uid\":\"4e014e03-0f47-406c-a738-bd66c645579a\",\"resourceVersion\":\"193\",\"creationTimestamp\":\"2024-01-21T11:27:24Z\",\"labels\":{\"component\":\"apiserver\",\"provider\":\"kubernetes\"},\"managedFields\":[{\"manager\":\"kube-apiserver\",\"operation\":\"Update\",\"apiVersion\":\"v1\",\"time\":\"2024-01-21T11:27:24Z\",\"fieldsType\":\"FieldsV1\",\"fieldsV1\":{\"f:metadata\":{\"f:labels\":{\".\":{},\"f:component\":{},\"f:provider\":{}}},\"f:spec\":{\"f:clusterIP\":{},\"f:internalTrafficPolicy\":{},\"f:ipFamilyPolicy\":{},\"f:ports\":{\".\":{},\"k:{\\\"port\\\":443,\\\"protocol\\\":\\\"TCP\\\"}\":{\".\":{},\"f:name\":{},\"f:port\":{},\"f:protocol\":{},\"f:targetPort\":{}}},\"f:sessionAffinity\":{},\"f:type\":{}}}}]},\"spec\":{\"ports\":[{\"name\":\"https\",\"protocol\":\"TCP\",\"port\":443,\"targetPort\":6443}],\"clusterIP\":\"fd00:10:96::1\",\"clusterIPs\":[\"fd00:10:96::1\"],\"type\":\"ClusterIP\",\"sessionAffinity\":\"None\",\"ipFamilies\":[\"IPv6\"],\"ipFamilyPolicy\":\"SingleStack\",\"internalTrafficPolicy\":\"Cluster\"},\"status\":{\"loadBalancer\":{}}}","ts":"2024-01-21T11:37:57Z"}
2024-01-21T11:37:57.243326269Z stdout F {"caller":"service_controller_reload.go:92","controller":"ServiceReconciler - reprocessAll","level":"debug","reprocessing service":"{\"kind\":\"Service\",\"apiVersion\":\"v1\",\"metadata\":{\"name\":\"kube-dns\",\"namespace\":\"kube-system\",\"uid\":\"4e2a2dd5-ec1f-418f-bca2-b94dff840582\",\"resourceVersion\":\"233\",\"creationTimestamp\":\"2024-01-21T11:27:25Z\",\"labels\":{\"k8s-app\":\"kube-dns\",\"kubernetes.io/cluster-service\":\"true\",\"kubernetes.io/name\":\"CoreDNS\"},\"annotations\":{\"prometheus.io/port\":\"9153\",\"prometheus.io/scrape\":\"true\"},\"managedFields\":[{\"manager\":\"kubeadm\",\"operation\":\"Update\",\"apiVersion\":\"v1\",\"time\":\"2024-01-21T11:27:25Z\",\"fieldsType\":\"FieldsV1\",\"fieldsV1\":{\"f:metadata\":{\"f:annotations\":{\".\":{},\"f:prometheus.io/port\":{},\"f:prometheus.io/scrape\":{}},\"f:labels\":{\".\":{},\"f:k8s-app\":{},\"f:kubernetes.io/cluster-service\":{},\"f:kubernetes.io/name\":{}}},\"f:spec\":{\"f:clusterIP\":{},\"f:internalTrafficPolicy\":{},\"f:ports\":{\".\":{},\"k:{\\\"port\\\":53,\\\"protocol\\\":\\\"TCP\\\"}\":{\".\":{},\"f:name\":{},\"f:port\":{},\"f:protocol\":{},\"f:targetPort\":{}},\"k:{\\\"port\\\":53,\\\"protocol\\\":\\\"UDP\\\"}\":{\".\":{},\"f:name\":{},\"f:port\":{},\"f:protocol\":{},\"f:targetPort\":{}},\"k:{\\\"port\\\":9153,\\\"protocol\\\":\\\"TCP\\\"}\":{\".\":{},\"f:name\":{},\"f:port\":{},\"f:protocol\":{},\"f:targetPort\":{}}},\"f:selector\":{},\"f:sessionAffinity\":{},\"f:type\":{}}}}]},\"spec\":{\"ports\":[{\"name\":\"dns\",\"protocol\":\"UDP\",\"port\":53,\"targetPort\":53},{\"name\":\"dns-tcp\",\"protocol\":\"TCP\",\"port\":53,\"targetPort\":53},{\"name\":\"metrics\",\"protocol\":\"TCP\",\"port\":9153,\"targetPort\":9153}],\"selector\":{\"k8s-app\":\"kube-dns\"},\"clusterIP\":\"fd00:10:96::a\",\"clusterIPs\":[\"fd00:10:96::a\"],\"type\":\"ClusterIP\",\"sessionAffinity\":\"None\",\"ipFamilies\":[\"IPv6\"],\"ipFamilyPolicy\":\"SingleStack\",\"internalTrafficPolicy\":\"Cluster\"},\"status\":{\"loadBalancer\":{}}}","ts":"2024-01-21T11:37:57Z"}
2024-01-21T11:37:57.243347258Z stdout F {"caller":"service_controller_reload.go:113","controller":"ServiceReconciler - reprocessAll","end reconcile":"metallbreload/reload","level":"info","ts":"2024-01-21T11:37:57Z"}
2024-01-21T11:37:57.243398493Z stdout F {"caller":"config_controller.go:58","controller":"ConfigReconciler","level":"info","start reconcile":"/kind-worker2","ts":"2024-01-21T11:37:57Z"}
2024-01-21T11:37:57.24432569Z stdout F {"caller":"config_controller.go:154","controller":"ConfigReconciler","level":"debug","rendered config":"(config.Config) {\n Peers: (map[string]*config.Peer) {\n },\n Pools: (*config.Pools)(0xc000e6a750)({\n  ByName: (map[string]*config.Pool) (len=1) {\n   (string) (len=7) \"l2-test\": (*config.Pool)(0xc0002c71f0)({\n    Name: (string) (len=7) \"l2-test\",\n    CIDR: ([]*net.IPNet) (len=8 cap=8) {\n     (*net.IPNet)(0xc000e6a390)(172.18.0.9/32),\n     (*net.IPNet)(0xc000e6a3c0)(172.18.0.10/31),\n     (*net.IPNet)(0xc000e6a3f0)(172.18.0.12/30),\n     (*net.IPNet)(0xc000e6a420)(172.18.0.16/30),\n     (*net.IPNet)(0xc000e6a600)(fc00:f853:ccd:e793::9/128),\n     (*net.IPNet)(0xc000e6a630)(fc00:f853:ccd:e793::a/127),\n     (*net.IPNet)(0xc000e6a660)(fc00:f853:ccd:e793::c/126),\n     (*net.IPNet)(0xc000e6a690)(fc00:f853:ccd:e793::10/126)\n    },\n    AvoidBuggyIPs: (bool) false,\n    AutoAssign: (bool) true,\n    BGPAdvertisements: ([]*config.BGPAdvertisement) <nil>,\n    L2Advertisements: ([]*config.L2Advertisement) (len=1 cap=1) {\n     (*config.L2Advertisement)(0xc000e6a720)({\n      Nodes: (map[string]bool) (len=3) {\n       (string) (len=18) \"kind-control-plane\": (bool) true,\n       (string) (len=11) \"kind-worker\": (bool) true,\n       (string) (len=12) \"kind-worker2\": (bool) true\n      },\n      Interfaces: ([]string) <nil>,\n      AllInterfaces: (bool) true\n     })\n    },\n    cidrsPerAddresses: (map[string][]*net.IPNet) (len=2) {\n     (string) (len=22) \"172.18.0.9-172.18.0.19\": ([]*net.IPNet) (len=4 cap=4) {\n      (*net.IPNet)(0xc000e6a390)(172.18.0.9/32),\n      (*net.IPNet)(0xc000e6a3c0)(172.18.0.10/31),\n      (*net.IPNet)(0xc000e6a3f0)(172.18.0.12/30),\n      (*net.IPNet)(0xc000e6a420)(172.18.0.16/30)\n     },\n     (string) (len=44) \"fc00:f853:ccd:e793::9-fc00:f853:ccd:e793::13\": ([]*net.IPNet) (len=4 cap=4) {\n      (*net.IPNet)(0xc000e6a600)(fc00:f853:ccd:e793::9/128),\n      (*net.IPNet)(0xc000e6a630)(fc00:f853:ccd:e793::a/127),\n      (*net.IPNet)(0xc000e6a660)(fc00:f853:ccd:e793::c/126),\n      (*net.IPNet)(0xc000e6a690)(fc00:f853:ccd:e793::10/126)\n     }\n    },\n    ServiceAllocations: (*config.ServiceAllocation)(<nil>)\n   })\n  },\n  ByNamespace: (map[string][]string) <nil>,\n  ByServiceSelector: ([]string) <nil>\n }),\n BFDProfiles: (map[string]*config.BFDProfile) {\n },\n BGPExtras: (string) \"\"\n}\n","ts":"2024-01-21T11:37:57Z"}
{"caller":"config_controller.go:58","controller":"ConfigReconciler","level":"info","start reconcile":"/kind-worker2","ts":"2024-01-21T11:38:57Z"}
2024-01-21T11:38:57.250991828Z stdout F {"caller":"bgp_controller.go:387","event":"nodeLabelsChanged","level":"info","msg":"Node labels changed, resyncing BGP peers","ts":"2024-01-21T11:38:57Z"}
2024-01-21T11:38:57.250996817Z stdout F {"caller":"speakerlist.go:274","level":"info","msg":"triggering discovery","op":"memberDiscovery","ts":"2024-01-21T11:38:57Z"}
2024-01-21T11:38:57.251001706Z stdout F {"caller":"config_controller.go:156","controller":"ConfigReconciler","event":"configuration did not change, ignoring","level":"debug","ts":"2024-01-21T11:38:57Z"}
2024-01-21T11:38:57.251005964Z stdout F {"caller":"node_controller.go:62","controller":"NodeReconciler","event":"force service reload","level":"info","ts":"2024-01-21T11:38:57Z"}
2024-01-21T11:38:57.251010012Z stdout F {"caller":"config_controller.go:157","controller":"ConfigReconciler","end reconcile":"/kind-worker2","level":"info","ts":"2024-01-21T11:38:57Z"}
2024-01-21T11:38:57.251013819Z stdout F {"caller":"node_controller.go:64","controller":"NodeReconciler","end reconcile":"/kind-worker2","level":"info","ts":"2024-01-21T11:38:57Z"}
2024-01-21T11:38:57.251038134Z stdout F {"caller":"main.go:297","event":"startUpdate","level":"debug","msg":"start of service update","ts":"2024-01-21T11:38:57Z"}
2024-01-21T11:38:57.251044305Z stdout F {"caller":"bgp_controller.go:170","event":"skipping should announce bgp","ips":["fc00:f853:ccd:e793::9"],"level":"debug","pool":"l2-test","protocol":"bgp","reason":"pool not matching my node","service":"l2-9268/external-local-lb","ts":"2024-01-21T11:38:57Z"}
2024-01-21T11:38:57.251048664Z stdout F {"caller":"layer2_controller.go:123","event":"shouldannounce","ips":["fc00:f853:ccd:e793::9"],"level":"debug","nodes":["kind-worker2","kind-control-plane","kind-worker"],"pool":"l2-test","protocol":"l2","service":"l2-9268/external-local-lb","ts":"2024-01-21T11:38:57Z"}
2024-01-21T11:38:57.251052811Z stdout F {"caller":"main.go:384","event":"serviceAnnounced","ips":["fc00:f853:ccd:e793::9"],"level":"info","msg":"service has IP, announcing","pool":"l2-test","protocol":"layer2","ts":"2024-01-21T11:38:57Z"}
2024-01-21T11:38:57.251057209Z stdout F {"caller":"main.go:346","event":"endUpdate","level":"debug","msg":"end of service update","ts":"2024-01-21T11:38:57Z"}
```

I can see from the kind logs that the speaker and controller components are working fine, it looks like the `GetSvcNode` function in the e2e code is not returning the announced node name correctly.

cyclinder commented:
```shell
Jan 23 03:18:27.[335](https://github.com/metallb/metallb/actions/runs/7620493254/job/20755437194?pr=2073#step:6:336): INFO: Waiting up to 2m0s for 1 pods to be running and ready: [external-local-lb-g277x]
Jan 23 03:18:29.341: INFO: Wanted all 1 pods to be running and ready. Result: true. Pods: [external-local-lb-g277x]
STEP: getting the advertising node @ 01/23/24 03:18:29.343
STEP: add the NodeExcludeBalancers label of the node @ 01/23/24 03:18:29.345
STEP: event.Message = announcing from node "kind-worker2" with protocol "layer2", event.Time = 1705979909000000000 
 @ 01/23/24 03:18:30.353
STEP: event.Message = announcing from node "kind-worker" with protocol "layer2", event.Time = 1705979909000000000
```

From the debug logs, this looks strange, the lastTimeStep is the same for different events, which causes `k8s.GetSvcNode` to return an incorrect node.

cyclinder commented:
@fedepaol All the tests have passed, We need the 1s sleep, Otherwise, the lastTimeStep is the same for different events, which causes k8s.GetSvcNode to return an incorrect node.

fedepaol commented:
> @fedepaol All the tests have passed, We need the 1s sleep, Otherwise, the lastTimeStep is the same for different events, which causes k8s.GetSvcNode to return an incorrect node.

Sorry if I insist, but I am probably missing something. Which GetSvcNode are you referring to? [This one](https://github.com/metallb/metallb/pull/2073/files#diff-07466f5529f653c060c48ccdc8852b5277ac753464747635fd346365270f9acaR322) (the one after the sleep) should work either way, as it's wrapped by eventually? And the previous one is before the sleep, so it won't be affected.

Also, if we decide the sleep is mandatory, we should add a big comment to explain why.

fedepaol commented:
> Please the [#2073 (comment)](https://github.com/metallb/metallb/pull/2073#issuecomment-1905436822), The second `GetSvcNode` actually gets two NodeAssigned events, but their times are the same, causing Eventually to always get the wrong node name. The reason for adding sleep before `AddLabelToNode` is to make sure that the second Event's time is later than the first Event's time. so we can get the correct node name.

Now I get it! So what you are actually delaying is adding the labels so the even has the right granularity.
This will change when we'll (eventually) add the status crds (one day!).
Just add a comment so it's clear, please.

fedepaol commented:
LGTM! Sending to merge queue

howardjohn commented:
I don't think this is a bug really, but fyi - https://github.com/metallb/metallb/issues/2274. This will break any `kind` users (among others)
